tests/test_0395-complex-type-arrays.py

```diff
@@ -0,0 +1,41 @@
+# BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE
+
+from __future__ import absolute_import
+
+import pytest  # noqa: F401
+import numpy as np  # noqa: F401
+import awkward as ak  # noqa: F401
+
+
+def test_UnmaskedArray():
+    content_float64 = ak.layout.NumpyArray(
+        np.array([0.25, 0.5, 3.5, 4.5, 5.5], dtype=np.float64)
+    )
+    array_float64 = ak.layout.UnmaskedArray(content_float64)
+    assert ak.to_list(array_float64) == [0.25, 0.5, 3.5, 4.5, 5.5]
+    assert str(ak.type(content_float64)) == "float64"
+    assert str(ak.type(ak.Array(content_float64))) == "5 * float64"
+    assert str(ak.type(array_float64)) == "?float64"
+    assert str(ak.type(ak.Array(array_float64))) == "5 * ?float64"
+
+    assert np.can_cast(np.float32, np.float64) == True
+    assert np.can_cast(np.float64, np.float32, "unsafe") == True
+    assert np.can_cast(np.float64, np.int8, "unsafe") == True
+    assert np.can_cast(np.float64, np.complex64, "unsafe") == True
+    assert np.can_cast(np.float64, np.complex128, "unsafe") == True
+    assert np.can_cast(np.complex64, np.float64, "unsafe") == True
+    assert np.can_cast(np.complex128, np.float64, "unsafe") == True
+
+    content_complex64 = ak.values_astype(content_float64, "complex64", highlevel=False)
+    array_complex64 = ak.layout.UnmaskedArray(content_complex64)
+    assert ak.to_list(array_complex64) == [0.25, 0.5, 3.5, 4.5, 5.5]
```

Comment:
@jpivarski - should it be: 
```python
[(0.25+0j), (0.5+0j), (3.5+0j), (4.5+0j), (5.5+0j)]
```

Reply:
ok, I think it's a feature :-)

Reply:
`ak.to_list` should turn arrays with complex dtype into Python `complex` numbers.

Some of the rules in `ak.to_list` convert to NumPy and call `np.ndarray.tolist`, which would make Python `complex` numbers if the NumPy `dtype` is a complex type, so either the NumPy `dtype` is coming out wrong or it's going through one of the other rules:

https://github.com/scikit-hep/awkward-1.0/blob/c171b744ea5cb4a68708fa7bd8bad64743747d34/src/awkward/operations/convert.py#L744-L772

Reply:
Ha... I forgot about this. I'm debugging it... I'm going to push one more commit to pretty-print complex NumpyArrays.

The following example:
```python
content_float64 = ak.layout.NumpyArray(
  np.array([0.25, 0.5, 3.5, 4.5, 5.5], dtype=np.float64)
)
content_complex64 = ak.values_astype(content_float64, "complex64", highlevel=False)
array_complex64 = ak.layout.UnmaskedArray(content_complex64)
```
it creates an
```python
<UnmaskedArray>
    <content><NumpyArray format="Zf" shape="5" data="0x 0000803e 00000000 0000003f 00000000 ... 00009040 00000000 0000b040 00000000" at="0x7f9979c44cf0"/></content>
</UnmaskedArray>
```
that is interpreted by `ak.to_list` as an `ak.layout.Content` or `ak.partition.PartitionedArray`. Then `ak.to_list` is iterating over the `bool`, `str`, `bytes`, or `numbers.Number` receiving complex numbers:
```python
to_list 
(0.25+0j)
 is bool, str, bytes, or numbers.Number
to_list 
(0.5+0j)
 is bool, str, bytes, or numbers.Number
to_list 
(3.5+0j)
 is bool, str, bytes, or numbers.Number
to_list 
(4.5+0j)
 is bool, str, bytes, or numbers.Number
to_list 
(5.5+0j)
 is bool, str, bytes, or numbers.Number
```
this results in:
```python
assert ak.to_list(content_complex64) == [0.25, 0.5, 3.5, 4.5, 5.5]
```
while the following works correctly:
```python
assert ak.to_list(ak.nplike.of(array_complex64).asarray(array_complex64)) == [(0.25+0.j), (0.5 +0.j), (3.5 +0.j), (4.5 +0.j), (5.5 +0.j)]
```
because `ak.to_list` is called on  `np.ndarray`:
```diff
to_list 
[0.25+0.j 0.5 +0.j 3.5 +0.j 4.5 +0.j 5.5 +0.j]
np.ndarray
```
Given an example of 
```python
    content_complex128 = ak.values_astype(content_float128, "complex128", highlevel=False)
    array_complex128 = ak.layout.UnmaskedArray(content_complex128)
    assert ak.to_list(content_complex128) == [0.25, 0.5, 3.5, 4.5, 5.5]
```
where
```
to_list 
<NumpyArray format="Zd" shape="5" data="0x 00000000 0000d03f 00000000 00000000 ... 00000000 00001640 00000000 00000000" at="0x7fe6f1e68ca0"/>
ak.layout.NumpyArray
```
`ak.to_list` is fine with `ak.layout.NumpyArray`:
```diff
to_list 
<NumpyArray format="Zd" shape="9" data="0x 9a999999 9999f13f 9a999999 9999b93f ... cdcccccc cccc2340 00000000 00000000" at="0x7f9979c37290"/>
ak.layout.NumpyArray
```

Reply:
This looks okay. It is not necessary to have pretty-printing of all NumPy dtypes in `NumpyArray::tostring_part`. If a dtype is not one of a small set of types we need for debugging, it prints the hex of the data.

The important thing is that `ak.to_list` produces Python `complex` objects, and if I'm reading the above correctly, it does. If that's the case, then you can resolve this conversation.

Reply:
Thanks! Yes, it does.